### DISCUSSION TITLE
In testcase, test-kens-transfer

### DISCUSSION THREAD

**Discussion by aaazzz1234** [ANLAB-KAIST/KENSv3#212] [Project Q&A 2. Transport Layer (TCP)] [answered]
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/138575713/fb0a7339-f81e-4605-8dae-3c9d5b17fe71)
위와 같이 test-kens-transfer의 첫번째 테스트케이스를 돌리는데 TestTransfer_Accept 부분의 read()에서 오류가 생깁니다.
제가 알기로 TEST_F(TestEnv_Any, TestTransfer_Connect_Send_Symmetric)의 TestTransfer_Accept 부분은 서버에서 진행하며 이 때 read()는 제가 구현한 read()가 아닌 기본적으로 있는 것을 사용하는 것으로 이해하였는데 왜 오류가 나는지 궁금합니다!

**Comment by Pusnow** [comment, marked as answer]
The -1 `read_byte` means `read()` is failed.
In KENS, if the simulation is terminated, `read()` returns -1.

It seems that your situation falls into this case. Since your data transfer attempt is failed and dismissed, the `read()` cannot return data on time, and the `read()` returns -1 at the end of the simulation.

**Reply by aaazzz1234** [reply]
넵 말씀하신 내용대로 write()에서 server로 data를 담은 packet을 전송하는 방식으로 구현하고 있습니다.
testcase 상에서는 128번 데이터를 전송하고 이후에 서버에서 다시 이를 read하는 방식으로 작동하는 것으로 이해하였는데
128번 write마다 데이터를 전송해주어야하는지 아니면 서버에서 ack-패킷을 보내는 것을 받을 때마다 1개씩 전송해야하는것인지 궁금합니다! 제가 맞게 이해하고 있는 것인가요??

**Reply by aaazzz1234** [reply]
또 test-kens-transfer의 첫번째 테스트케이스에서 read와 write 모두 구현해야지 정상적으로 작동하는 것인지 궁금합니다!
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/138575713/f32c27f3-bca9-4d5b-ae4b-45cfff6d8557)
제가 구현한 좌측과 실제 답안인 우측을 비교하면 서버로부터 답변이 오지 않아 클라이언트의 추가적인 패킷전송이 안되는 상황인 것으로 이해했는데 차이점이 클라이언트의 포트번호인 것 같다고 생각합니다. 혹시 클라이언트의 포트번호를 임의로 지정하면 안되는지, 만약 그렇다면 어떤 방식으로 지정해주어야하는지 여쭤보고 싶습니다...ㅠㅠ

바쁘신 와중에 답변 주셔서 진심으로 감사드립니다😊

**Reply by Pusnow** [reply]
> 128번 write마다 데이터를 전송해주어야하는지 아니면 서버에서 ack-패킷을 보내는 것을 받을 때마다 1개씩 전송해야하는것인지

I recommend sending packets on each write() calls if rwnd is allowed.

> 또 test-kens-transfer의 첫번째 테스트케이스에서 read와 write 모두 구현해야지 정상적으로 작동하는 것인지 궁금합니다!

Which test case? Since the other peer is TA's implementation, you probably can pass some test cases even though you only implement either read or write calls.

> 혹시 클라이언트의 포트번호를 임의로 지정하면 안되는지, 만약 그렇다면 어떤 방식으로 지정해주어야하는지 여쭤보고 싶습니다...ㅠㅠ

You can use random port numbers as long as they are not conflicted (but port number zero should not be used).
